### ISSUE TITLE
Self-hosted runner stuck on "Waiting for a runner to pick up this job..." in multi-step jobs

### ISSUE THREAD

**Issue by joelnwalkley** [actions/runner#3609]
**Describe the bug**
Using a self-hosted runner.
When a GitHub action has multiple steps the first will run successfully but then the subsequent step is stuck in queued status with a logging message "Waiting for a runner to pick up this job..."

I am able to get the stuck jobs to start by either cancelling and re-running the job via the GitHub UI or by restarting the GitHub Runner service within our EC2. In both instances the job immediately picks up and runs successfully. 

**To Reproduce**
1. Use self-hosted GitHub Runner
2. Use a multi-step action
3. Initiate the action
4. Observe that the 2nd step is stuck in queued status.

**Expected behavior**
The runner should pick up the 2nd (and following steps)

## Runner Version and Platform
v2.321.0 Windowsx64

We noticed that our machine auto-updated with this version on November 27 and then our CI runs the following week started to have this problem.

OS of the machine running the runner?
Windows

## What's not working?
"Waiting for a runner to pick up this job..." for up to hours.

## Job Log Output
N/A the runner does not get to job output, it is stuck in queue.

**Comment by daniel-v-nanok** [comment]
I am also being affected by this, tried everything, added disableUpdate to no avail

**Comment by richardgetz** [comment]
Same here. Manually stopping and starting the service makes it move onto the next job. Obviously this is not ideal.

**Comment by TylerWilliamson** [comment]
Also affected by this issue v2.321.0 on Debian 12 / amd64. We are able to workaround this issue by rebooting the runner.

**Comment by richardgetz** [comment]
I tested downgrading to v2.320.0 and encountered the same issue.

**Comment by canmustu** [comment]
Same. I tried everything that I can. Nothing changed.

**Comment by amalhanaja** [comment]
Same here. is there any work around for this issue?

**Comment by connor-27** [comment]
I've tried almost everything I could do... but it's not working out well.

**Comment by rohitkhatri** [comment]
I'm also facing the same issue, have to stop the service and start again then only the next stage starts

**Comment by SoftradixAD** [comment]
it's not working out well.

**Comment by HoaiTTT-Kozocom** [comment]
I'm also facing the same issue, have to stop the service and start again then only the next stage starts

**Comment by kemalgoekhan** [comment]
Windows and Linux both have same issue.

**Comment by enginelesscc** [comment]
Same issue here, and my workaround is to rerun & immediately cancel some old action.
This "revives" stuck jobs, but new jobs endup with the same problem again.

**Comment by dexpert** [comment]
same issue

**Comment by NekiHrvoje** [comment]
For those using action `machulav/ec2-github-runner` which does not use systemd service you need to kill process `/actions-runner/run-helper.sh` and start it again.

**Comment by netaviator** [comment]
Same issue here. Tried reinstalling and downgrading the runner without success.

**Comment by lokesh755** [comment]
Could you provide run url which got stuck waiting for runner? That'll help us debug.

**Comment by m-iwanicki** [comment]
I'm having the same problem, this run got stuck for whole day: https://github.com/Dasharo/meta-dts/actions/runs/12163220462. Had to restart runner so workflow would continue.
Weirdly this one https://github.com/Dasharo/meta-dts/actions/runs/12180581391 got stuck waiting on `Run DTS tests` but cleanup started normally after job failed.

**Comment by jgrasett** [comment]
We've been seeing the same thing here for 3 days now. First 2 jobs run, then waiting...

We are seeing this in multiple repositories, though some of them work fine. Nothing in our workflows has changed in 6 months.

Ubuntu 22.04 EC2 instances in AWS.
using:  machulav/ec2-github-runner@v2
also tried: machulav/ec2-github-runner@v2.3.7 but no changes to issue.

**Comment by canmustu** [comment]
@lokesh755 same here for self-hosted private repos.

https://github.com/RapiderAI/turplus--20197491-0906-4f36-8ce4-c5fa9d1e22d7/actions/runs/12179682401

**Comment by harricross** [comment]
This caused issues for Merck (P+)'s ARC cluster, where runners could not de-register themselves to cleanly tear down, and this caused a lot of issues on their Kubernetes cluster. [Ticket here](https://github.zendesk.com/agent/tickets/3127212)

**Comment by Meigara-Juma** [comment]
> Thanks for reporting the issue. We've identified the root cause, which appears to be linked to a feature flag that was enabled two days ago. We've temporarily disabled the feature flag, which should resolve the issue. If you continue to experience similar problems, please let us know.

Has anyone tried it and got it working normally again? Because I have done a workaround by changing multiple jobs to a single job in my workflow

**Comment by canmustu** [comment]
> > Thanks for reporting the issue. We've identified the root cause, which appears to be linked to a feature flag that was enabled two days ago. We've temporarily disabled the feature flag, which should resolve the issue. If you continue to experience similar problems, please let us know.
> 
> Has anyone tried it and got it working normally again? Because I have done a workaround by changing multiple jobs to a single job in my workflow

It is fixed on my workflows. No issue at all after @lokesh755 's last message for me.

**Comment by netaviator** [comment]
Can confirm! 👍🏿

**Comment by SoftradixAD** [comment]
I'm still facing this

**Comment by rohitkhatri** [comment]
Fixed for me as well

@SoftradixAD can you try stopping and starting the service again, that worked for me

**Comment by SoftradixAD** [comment]
Okay let me try, @rohitkhatri sir

**Comment by NorseGaud** [comment]
Is https://github.com/orgs/community/discussions/146348 related?

I tried late yesterday after the feature flag was disabled and I still see the problem.

**Comment by daniel-v-nanok** [comment]
> Thanks for reporting the issue. We've identified the root cause, which appears to be linked to a feature flag that was enabled two days ago. We've temporarily disabled the feature flag, which should resolve the issue. If you continue to experience similar problems, please let us know.

Thank you, it is working correctly

**Comment by connor-27** [comment]
> Thanks for reporting the issue. We've identified the root cause, which appears to be linked to a feature flag that was enabled two days ago. We've temporarily disabled the feature flag, which should resolve the issue. If you continue to experience similar problems, please let us know.

Thank you~! it is nice

**Comment by cgerace** [comment]
I'm still experiencing this issue, is anyone else?

**Comment by NorseGaud** [comment]
I see https://github.com/orgs/community/discussions/146348 still. Github support has been unhelpful so far.

**Comment by heikkis** [comment]
No issues on our side any more after the feature flag was disabled (https://github.com/actions/runner/issues/3609#issuecomment-2520606406) . Running multi-step jobs with a dynamically self hosted runner.

**Comment by Oladejiraji** [comment]
Still facing this issue. Any solutions ?

**Comment by cgerace** [comment]
@Oladejiraji I had to ssh into the virtual machine that was hosting our runner (EC2) and remove all unused docker images, I believe i ran `docker image prune -a`. That appears to have done the trick.

**Comment by myusuf3** [comment]
I am still seeing this and spun up this self hosted runner a few days ago.

**Comment by myusuf3** [comment]
> I am still seeing this and spun up this self hosted runner a few days ago.

I have stopped and started the service multiple times to get it to pick it up.

**Comment by biswajitm2024** [comment]
I have also been facing the same issues since January 9. I tried it several times; sudo ./svc.sh stop and sudo ./svc.sh start are not working for me. My version is v2.321.0.

**Comment by Flodax** [comment]
@lokesh755 which feature flag has to be disabled and how can we do it? We are facing the same issue

**Comment by bogo-bohr** [comment]
> @lokesh755 which feature flag has to be disabled and how can we do it? We are facing the same issue

I guess it was on github.com the error was. So when @lokesh755 posted that the problem was fixed (feature flag disabled) maybe you only needed to retrigger a new workflow or do whatever cleaning for it to be running again. This is my guess based on all these people just posting that it now works.

However, I am also stuck with the same error still even after rebooting my local runner machine and retriggering the workflows. So I am also curious on what else it could be or what should be cleaned up for it to run again (using Windows 10 Pro - v 22H2 and runner v 2.321.0).

**Comment by NorseGaud** [comment]
Likely related: https://github.com/actions/runner/issues/3621 and https://github.com/orgs/community/discussions/146348

**Comment by bogo-bohr** [comment]
It fixed my problem to re-configure the local machine like this:
1. delete the files .runner, .credentials and .credentials_rsaparams
2. start config.cmd and answer the questions asked (see below to get temp register token)
3. get register token by typing this into a new command prompt substitute ORGANIZATION and REPO:
      gh api --method POST -H "Accept: application/vnd.github+json" /repos/ORGANIZATION/REPO/actions/runners/registration-token
4. start run.cmd

Maybe this can help someone.

**Comment by daraccrafter** [comment]
I am still facing this issue. Any update on this?

**Comment by cafecafe** [comment]
I'm having a very similar issue where every 2nd task of the runner is stuck in "queued" status.

The cause of this is the runner not properly cleaning up state between operations

Stopping the runner in systemctl, running the following

`rm -rf _work/*
rm -rf _diag/*.json`

And starting it again makes it pick up the task immediately. I thought this might be related, it gets stuck on the job level in my case, not on the step level.

**Comment by cafecafe** [comment]
I would like to update that my case is no longer relevant. The issue was a conflicting runner.

Some details:
One runner was newly added (v 2.322). Another one was an older runner. I saw some 2.322.0 binaries in the folder, so I assumed that it was auto-updated. Apparently, it conflicted. Removing the older runner completely and setting up the 2.322.0 from scratch fixed the issue. Both runners started to pick up the jobs immediately.

Please disregard my case.

**Comment by cateyelow** [comment]
In my case, it resolved by setting the label as simple digits. if you have only one self-runner , set the label as `self-hosted`


```
jobs:
  deploy-server-a:
    runs-on: [self-hosted, Windows, X64, 1]

  deploy-server-b:
    runs-on: [self-hosted, Windows, X64, 2]

```

**Comment by LookedPath** [comment]
I've been having the exact same issue for a few days, I even tried deleting all runners and setting just one up from scratch but the issue remains.

**Comment by yiwansk** [comment]
Can it be assigned higher priority? Having to restart tasks manually is a game breaking bug for a CI system.

**Comment by JonathanCroenen** [comment]
Started having the same issue, my runners are issue persists from v2.320.0 - v2.322.0 after manually updating them and retrying (auto update also doesn't ever pop up for me...). Runners will randomly stop picking up jobs, cancelling the workflows and restarting them just fixes it, so it's the I don't see how it can be a config issue if they just pick them up again without any changes.

**Comment by daraccrafter** [comment]
> Started having the same issue, my runners are issue persists from v2.320.0 - v2.322.0 after manually updating them and retrying (auto update also doesn't ever pop up for me...). Runners will randomly stop picking up jobs, cancelling the workflows and restarting them just fixes it, so it's the I don't see how it can be a config issue if they just pick them up again without any changes.

I agree one day they pick up jobs normally, other day they just dont.

**Comment by cinemast** [comment]
I am also affected by this. Restarting the service works, but I can't restart the services the whole day for all runners :D

**Comment by oxisto** [comment]
Same for us... worked fine until more or less yesterday and the day before, now the queue gets "stuck" every 5 executed runs. interestingly enough at some point (10m+) the jobs are getting picked up... But until then the actions runner is just idle 🤷

**Comment by zzlk** [comment]
This is also affecting my CI pipelines. restarting the job or restarting the runner will make it pick up instantly, otherwise it just sits there.

**Comment by nanolookc** [comment]
Same, for last week always should manually restart runner or restart job in actions panel

**Comment by derwin12** [comment]
Same here. I get the success message but never goes back to listening for jobs. Not running as a service btw

**Comment by pacosw1** [comment]
Same here. It's not stuck forever but it used to take seconds, now it takes up to 5-10min to pick up the jobs queued. My server is idle and the worker running

**Comment by derwin12** [comment]
> Same here. I get the success message but never goes back to listening for jobs. Not running as a service btw

Cant run as a service as I need the GUI. Is there a debug option or something on ./run.cmd that might be helpful to trace?

![Image](https://github.com/user-attachments/assets/1dc65929-05ce-4f68-8722-0a4c47076c33)

[Worker_20250308-153447-utc.log](https://github.com/user-attachments/files/19144497/Worker_20250308-153447-utc.log)

No active jobs..
![Image](https://github.com/user-attachments/assets/e73aa981-c88c-47c4-9820-a06552182517)

Stuck in the queue

![Image](https://github.com/user-attachments/assets/ebb72118-f6cc-4c2e-aa7e-6e42dc848dfe)

**Comment by derwin12** [comment]
Duplicate of #3478

**Comment by xorth** [comment]
+1, impacted over here too.

**Comment by zzzhr1990** [comment]
It seems that the issue has occurred again, and even restarting the service doesn't work now.

**Comment by NevenD** [comment]
+1, Same here as well.

**Comment by svenbrill** [comment]
+1

**Comment by kioz-wang** [comment]
+1

**Comment by kioz-wang** [comment]
+1

**Comment by TingluoHuang** [comment]
probably related to https://www.githubstatus.com/incidents/m7vl0x8k3j9c

**Comment by abrinkman** [comment]
+1, same here

**Comment by derwin12** [comment]
I have had zero issues since I went to the cmd line and set my git credentials. I surmise that there were 2 accounts and the listener didnt know which one to use.

**Comment by artyomsv** [comment]
+1

**Comment by msneijders** [comment]
+1

**Comment by daraccrafter** [comment]
+1

**Comment by sadewa25** [comment]
+1

**Comment by Ooooze** [comment]
+1

**Comment by AndriyNakonechnyy** [comment]
+1

**Comment by berkaycit-rotatelab** [comment]
+1

**Comment by abhi-iac** [comment]
+1 Still encountering issue. Any workaround?

**Comment by innocentinnox** [comment]
same issue with my selfhosted runner on debian

**Comment by TingluoHuang** [comment]
Can anyone share their workflow job URL that's waiting for runner, and the name of the self-hosted runner that suppose to pick up the job? 🙇

**Comment by theinspire** [comment]
+1

**Comment by 295722767** [comment]
+1 Same here

**Comment by YanzhaoW** [comment]
+1 same here. 

I have to either restart the service or cancel the workflow and rerun it.

**Comment by TavioDot** [comment]
+1

**Comment by jelsinor** [comment]
+1

**Comment by derwin12** [comment]
> Can anyone share their workflow job URL that's waiting for runner, and the name of the self-hosted runner that suppose to pick up the job? 🙇

It picks up the first then nothing..
Is there an option to ./run.cmd to verbose?

![Image](https://github.com/user-attachments/assets/8445cf15-b237-480e-908a-d5e1c359fe2f)
After pressing ctrl_c twice it breaks out with terminate batch job (y/n)
after terminating and restarting it picks up just fine - for 1 job atleast.

**Comment by nhoss2** [comment]
Getting this quite a bit now on a job that gets queued on a cron and has a max 1 concurrent setting

**Comment by VSilantyev** [comment]
+1

**Comment by 5priyanka** [comment]
I have the same issue. I uninstalled the services, reinstalled them, registered the runner again, and also changed the labels, but no luck. Every time, I need to restart for it to pick up the job. This has been happening for the last two weeks; otherwise, it has worked well for three years. Do we have a solution? I’m not keeping track of who made changes, and I have to restart the service. Thank you.

**Comment by arjunsbabu43** [comment]
+1

**Comment by VSilantyev** [comment]
When I encounter this issue, I see errors like this, about _temp cleanup in _diag job logs:
```
tail -n 100 -f _diag/Worker_20250313-075501-utc.log

[2025-03-13 07:51:41Z INFO TempDirectoryManager] Cleaning runner temp folder: /opt/actions-runner/_work/_temp
[2025-03-13 07:51:41Z ERR  TempDirectoryManager] System.IO.IOException: Access to the path '/opt/actions-runner/_work/_temp/_github_home/.npm/_cacache/content-v2/sha512/a7/6a' is denied.
   at System.IO.FileSystem.RemoveEmptyDirectory(String fullPath, Boolean topLevel, Boolean throwWhenNotEmpty)
   at System.IO.DirectoryInfo.Delete()
   at GitHub.Runner.Sdk.IOUtil.DeleteDirectory(String path, Boolean contentsOnly, Boolean continueOnContentDeleteError, CancellationToken cancellationToken)
   at GitHub.Runner.Worker.TempDirectoryManager.CleanupTempDirectory()

```

I've cleaned _temp:
```
sudo ./svc.sh stop
sudo rm -Rf /opt/actions-runner/_work/_temp/*
sudo ./svc.sh start
```

And runner started to pickup jobs immediately.

**Comment by giellamo** [comment]
Here also.

**Comment by msneijders** [comment]
@TingluoHuang
> Can anyone share their workflow job URL that's waiting for runner, and the name of the self-hosted runner that suppose to pick up the job? 🙇

https://github.com/florisoft/ontwikkel/actions/runs/13833899933/job/38704282724

**Comment by msneijders** [comment]
@TingluoHuang
> Can anyone share their workflow job URL that's waiting for runner, and the name of the self-hosted runner that suppose to pick up the job? 🙇

https://github.com/florisoft/ontwikkel/actions/runs/13837481411/job/38716106220

name of self-hosted runner is 'BUILDAGENT1-ontwikkel' (also as a label added)

**Comment by yaananth** [comment]
We apologize for the continued issues with self-hosted runners not picking up jobs. We identified a longstanding problem related to runner connections, specifically in determining when runners are online versus offline. A recent rollout unintentionally amplified this issue, causing more frequent disruptions.

We’ve since rolled back that change, reproduced the issue internally, implemented a fix, and confirmed that the problem no longer occurs. The corrected update has now been rolled out again.

If you’re still experiencing this issue, please report it and let us know. Any jobs that are currently pending will need to be canceled and retried or triggered again.

Thank you for your patience!

**Comment by mukhamux** [comment]
hey @yaananth 
> The corrected update has now been rolled out again.

what version I should use to avoid this error?

**Comment by yaananth** [comment]
> hey @yaananth 
> > The corrected update has now been rolled out again.
> 
> what version I should use to avoid this error? It is still stuck on 2.322.0

Shouldn't be related to runner version, have a run URL? Providing run URL helps us to investigate closely, as the issue could happen due to various reasons

**Comment by mukhamux** [comment]
> Providing run URL helps us to investigate closely, as the issue could happen due to various reasons

this one?
https://github.com/mukhamux/parad/actions/runs/13844032966/job/38738457946

**Comment by Banck** [comment]
+1

**Comment by yaananth** [comment]
> > Providing run URL helps us to investigate closely, as the issue could happen due to various reasons
> 
> this one? https://github.com/mukhamux/parad/actions/runs/13844032966/job/38738457946

This is a configuration issue looks like, the job is requesting `3800529-mukhamux` as label, but runner doesn't have that label, it seems like you are using runner name to request as label. Please add appropriate labels to runner. https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/using-labels-with-self-hosted-runners#creating-a-custom-label

**Comment by Banck** [comment]
@yaananth 
https://github.com/Olimpcom-organization/olimpcom-ios/actions/runs/13833250743

**Comment by yaananth** [comment]
@Banck could you file a new issue with [runner diagnostic logs](https://github.com/actions/toolkit/blob/main/docs/action-debugging.md#runner-diagnostic-logs), looking at that run, it's not related to the issue we fixed/looked here, for some reason the runner isn't communicting back in your looks like, please create a new one for this with logs, thank you!

**Comment by giovegaq** [comment]
> We apologize for the continued issues with self-hosted runners not picking up jobs. We identified a longstanding problem related to runner connections, specifically in determining when runners are online versus offline. A recent rollout unintentionally amplified this issue, causing more frequent disruptions.
> 
> We’ve since rolled back that change, reproduced the issue internally, implemented a fix, and confirmed that the problem no longer occurs. The corrected update has now been rolled out again.
> 
> If you’re still experiencing this issue, please report it and let us know. Any jobs that are currently pending will need to be canceled and retried or triggered again.
> 
> Thank you for your patience!

I had the same problem today, I saw that removing the label and adding it again inmediately the runner start picking jobs

**Comment by giovegaq** [comment]
> We apologize for the continued issues with self-hosted runners not picking up jobs. We identified a longstanding problem related to runner connections, specifically in determining when runners are online versus offline. A recent rollout unintentionally amplified this issue, causing more frequent disruptions.
> 
> We’ve since rolled back that change, reproduced the issue internally, implemented a fix, and confirmed that the problem no longer occurs. The corrected update has now been rolled out again.
> 
> If you’re still experiencing this issue, please report it and let us know. Any jobs that are currently pending will need to be canceled and retried or triggered again.
> 
> Thank you for your patience!

I'm getting the issue here (the runner looks like idle):
https://github.com/sorenson-eng/vri-ios/actions/runs/13861680800/job/38791488661

Waiting for a runner to pick up this job...

**Comment by spdesai** [comment]
We are seeing the same issue as of today.

**Comment by Ryangr0** [comment]
Just happened to us as well for the Nth time, which caused me to look around and stumbled upon this thread. Been noticing this more the past few weeks, but I don't know if it's related to the changes we're making in our runners setup (We're implementing https://github.com/actions/actions-runner-controller), or if it's related to something upstream.

**Comment by DavidPriceNBS** [comment]
This is massively blocking us, I'm experiencing this issue when a workflow does the following

Job Requiring Approval -> Parallel Jobs after approval e.g 

![Image](https://github.com/user-attachments/assets/307312a8-3708-4383-9fec-c4af58958ef2)

The "Release" job requires approval, then the subsequent "fe-sync-storage" job runs on the self-hosted (in Azure Container Apps Jobs) runners and should trigger those and execute - however they poll infinitely. There are no issues with configuration or labels, as if I do the exact same thing but without that pre-approval step they execute fine, see below. 

(Note - The actual workflow executing the sync is the same for both and being called as a reusable workflow which is why I know the functionality works)

![Image](https://github.com/user-attachments/assets/d34968a5-7ffe-4e43-9610-6f80a2173a85)

It simply never triggers now, we're unable to proceed with using the runner. I've tried redeploying the container app job, rebuilding the image with the latest 3.22.0 version, it just fails to ever trigger.

**Comment by okmanideep** [comment]
Had to remove the runner from the organisation. Delete the runner directory on the machine. And re-add the runner (re-download + configure) to get it to start picking up jobs again. Not sure if this will last. But if someone wants to get their runners working again desperately, it's worth a shot.

**NOTE**: make sure you save the `.env` file before deleting the `actions-runner` directory

**Comment by mklauber** [comment]
I"m having issues with this, using the Helm chart.  Example waiting action here: https://github.com/DirectEmployers/microsites-states/actions/runs/14106168137/job/39513127407

**Comment by froblesmartin** [comment]
> I'm having issues with this, using the Helm chart. Example waiting action here: [DirectEmployers/microsites-states/actions/runs/14106168137/job/39513127407](https://github.com/DirectEmployers/microsites-states/actions/runs/14106168137/job/39513127407)

Same thing here, we have GitHub Runners with the official Helm chart running in K8s, and somehow the listener pods receive the request to spin up a runner for a workflow job, but it does not happen, or happens anytime (even hours) later.

Only us having this issue? 🤔

**Comment by PaulsenDanielMeta** [comment]
I am also seeing this issue on many of my run attempts.  Running 8 shards utilizing 2 workers.  Always fails picking up the last shard.

**Comment by SwapnilSoni1999** [comment]
![Image](https://github.com/user-attachments/assets/2caa2580-7814-4e7a-a5ff-6025732907ce)

something is terribly wrong here

**Comment by zarko-a** [comment]
We are experiencing the same issue with ARC. Worked fine until recently, now we get jobs stuck in queued state for extended periods of time.

Anyone found a solution or a workaround?

**Comment by SwapnilSoni1999** [comment]
> We are experiencing the same issue with ARC. Worked fine until recently, now we get jobs stuck in queued state for extended periods of time.
> 
> Anyone found a solution or a workaround?

I restarted the service using systemctl and after a while it picked up (~30mins)

**Comment by zarko-a** [comment]
Thanks, but still seems it's kinda broken. I'll open a ticket with GH support and report back if we get anything useful.

**Comment by NikulinIgor** [comment]
Same issue.
Using a Windows self-hosted runner.
Everything was working fine until today, and no changes were made to the workflow.
Starting from the second step, the workflow hangs at:
"Waiting for a runner to pick up this job..."

However, if I stop the workflow and re-run the stuck job, it completes successfully.

**Comment by zarko-a** [comment]
We found a workaround, at least it helped for now. Uninstalling scale sets **and** the actions runner controller completely from the cluster, then redeploying using the same helm values as before made it work. Must be a bug with the controller since we use ArgoCD and the code this is deployed from hasn't changed at all.

**Comment by surgiie** [comment]
> We found a workaround, at least it helped for now. Uninstalling scale sets **and** the actions runner controller completely from the cluster, then redeploying using the same helm values as before made it work. Must be a bug with the controller since we use ArgoCD and the code this is deployed from hasn't changed at all.

This seems to have fixed this for us (for now). We've been having this issue randomly crop up frequently. Seems like a major bug that needs more attention.

**Comment by jakerobb** [comment]
My team has been encountering this issue for the past couple weeks on our private repos with self-hosted runners. Here's a URL for one that's hanging as I type: https://github.com/BusinessWire/bw-accounts/actions/runs/15567325350/job/43835268858

**Comment by zarko-a** [comment]
Is there anyone from Github actually reading these comments? 

I opened a ticket with them 3 business days ago and haven't heard anything, and that was through a paid account.

**Comment by surgiie** [comment]
> Is there anyone from Github actually reading these comments?
> 
> I opened a ticket with them 3 business days ago and haven't heard anything, and that was through a paid account.

@zarko-a 
Curious if you have heard anything back? 

The issue keeps coming back for us every morning. Specifically we have alot of scheduled builds that run first thing in the morning, those seem to get stuck and then hold up any other builds from running. We've been canceling the builds manually and that gets things flowing again but its not an ideal fix. At this point, we're considering an alternative vendor.

**Comment by zarko-a** [comment]
I haven't heard anything since. It seems they aren't interested in providing support. My guess is they want everyone to use their managed runners. Self-hosted runners get very little love.

**Comment by jakerobb** [comment]
Plenty of us would love to use managed runners but cannot due to contractual and/or security restrictions.

**Comment by surgiie** [comment]
> Plenty of us would love to use managed runners but cannot due to contractual and/or security restrictions.

Yeah same. We need to use our runners for security restrictions and our workers need to talk to things within our internal network.

**Comment by thboop** [comment]
Hey folks, I did want to say we are actively monitoring this repo.

There's quite a few escalations that have come for self hosted runners recently, including some from this thread. 

If you are using ARC, and you are seeing jobs pending, the most common cause is "failed pods". These pods take up a scaling slot, so if you have a cluster with 1 failed pod, and 1 job assigned, that job will never start and you will see the queued behavior. ARC users are responsible for monitoring these failed pods and ensuring they are cleaned up in a timely manner. You can see a community post on how to clean those up [here](https://github.com/orgs/community/discussions/160444#discussioncomment-10397035). Last week a server incident caused more pods to fail, causing more users to end up in this state.

They show up in metrics like

```
2025-06-01T00:00:00Z	INFO	EphemeralRunnerSet	Ephemeral runner counts	{"version": "0.11.0", "ephemeralrunnerset": {"name":"redacted","namespace":"redacted"}, "pending": 0, "running": 1, "finished": 1, "failed": 30, "deleting": 0}
```


This is currently expected behavior. When a workflow fails the runners enter a failed state where they are not automatically removed. This allows customer to review the logs from their runners to determine the cause of the failure and resolve that. The customer can manually remove the failed pods once they've completed any investigations. 

I've also bubbled up feedback to the team that just logging failure reasons and automatically healing those failed runners would be a much better experience, I can't promise anything but I hope we can improve that experience.


If you are seeing a non ARC issue, please create a separate ticket, describing the errors you are seeing, provide a workflow run url, and the [runner diagnostic logs ](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/monitoring-and-troubleshooting-self-hosted-runners#reviewing-the-self-hosted-runner-application-log-files)

**Comment by zarko-a** [comment]
@thboop Thanks for getting back to us. We are indeed using ARC, but unfortunately, I don't think the issue we encountered aligns with the scenario you described with failed pods. In our case there were 0 pods (failed or running) in our runner namespace yet nothing was getting scheduled. Everything was just queued for over 10 hours, until we reinstalled ARC controller. 

If it helps, here is a ticket number 3461184. There I've included some screenshots.

**Comment by thboop** [comment]
Hey @zarko-a, the pods would not appear as running, they are failed which may be causing some confusion.


Looking at your ticket I would strongly encourage you to double check your controller logs for a "EphemeralRunnerSet" and follow the guidance [here](https://github.com/orgs/community/discussions/160444#discussioncomment-10397035) if your failed number is anything other then 0. It sounds like it was mitigated by a redeploy of the cluster on your end.

It would look something like
```
"failed": x, "deleting": 0}
```

**Comment by thboop** [comment]
You could also follow the [steps](https://github.com/actions/runner/issues/3609#issuecomment-2956495536) here to fully re-setup the cluster

The good news is that a [fix](https://github.com/actions/actions-runner-controller/issues/4091#issuecomment-2958375837) for this manual cleanup should be coming in the next version of ARC, sorry for the churn here

**Comment by surgiie** [comment]
> You could also follow the [steps](https://github.com/actions/runner/issues/3609#issuecomment-2956495536) here to fully re-setup the cluster
> 
> The good news is that a [fix](https://github.com/actions/actions-runner-controller/issues/4091#issuecomment-2958375837) for this manual cleanup should be coming in the next version of ARC, sorry for the churn here

For our setup deleting failed ephemeral runners only worked for a while as long as the runners were marked with status `Failed` but we started getting ephemeral runners that were just completely stuck in a `<none>` phase which we confirmed with 

`kubectl get EphemeralRunner -n github-actions --no-headers -o 'custom-columns=:metadata.name,:status.phase'`. 

I assume that `<none>`  is either `pending` or `queued`, im not entirely sure how to confirm.

**Comment by zarko-a** [comment]
Thanks @thboop . I don't remember if checked EphemeralRunners resource when the issue was manifesting. I'll keep this in mind and check it next time when it happens.

**Comment by kyrylomiro** [comment]
Same kind of issue happens to us, some of our runners group are particularly affected, and they have `<none>` status

**Comment by surgiie** [comment]
For us, we updated to latest ARC version, and segmented some of our scheduled builds out and things were working great for about a week, no stuck builds but this morning we started to experience this issue again:

A few builds stuck for about 20 minutes.  We had 6 EphemeralRunners in the "Running" state, but no pods started.  Deleting the EphemeralRunners triggered new runners to start, which in turn started new pods that processed immediately.

Some type of bug is still at play in latest version where things are not being  started and get stuck.  


There is a thread/issue on the ARC repo as well about this: https://github.com/actions/actions-runner-controller/issues/4121

**Comment by elaigor** [comment]
So, after the last updates, we are not having any EphemeralRunner with status Failed, BUT now, all those runners that are continuing are stuck with status ready = false. 
After cleaning these EphemeralRunners, the problem with "Waiting for a runner to pick up this job" disappears, but until the next failures. 
```
status:
  failures:
    b571c315-......-db8b2a2c340b: '2025-07-01T08:03:00Z'
  phase: Running
  ready: false
```

**Comment by surgiie** [comment]
> So, after the last updates, we are not having any EphemeralRunner with status Failed, BUT now, all those runners that are continuing are stuck with status ready = false. After cleaning these EphemeralRunners, the problem with "Waiting for a runner to pick up this job" disappears, but until the next failures.
> 
> ```
> status:
>   failures:
>     b571c315-......-db8b2a2c340b: '2025-07-01T08:03:00Z'
>   phase: Running
>   ready: false
> ```

Is this on the latest 0.12.1 version? https://github.com/actions/actions-runner-controller/releases/tag/gha-runner-scale-set-0.12.1

**Comment by elaigor** [comment]
> Is this on the latest 0.12.1 version? https://github.com/actions/actions-runner-controller/releases/tag/gha-runner-scale-set-0.12.1

No. It's from 0.12.0. 
I will try to update to 0.12.1 in a few days. 
Thanks

**Comment by elaigor** [comment]
Can confirm that 0.12.1 fixes the problem with EphemeralRunner in failed status, and for 24h we haven't had any "Waiting for a runner to pick up this job...". 
Thanks!

**Comment by Patel-Aakash** [comment]
I'm also facing the same issue on windows self hosted runner.

**Comment by sergeishestakov-gh** [comment]
is it any solution for self-hosted vm runners ? 
same issue on 2.321.0 version

**Comment by SwapnilSoni1999** [comment]
> is it any solution for self-hosted vm runners ? same issue on 2.321.0 version

Restart the service and it'll be back online

**Comment by sergeishestakov-gh** [comment]
> Restart the service and it'll be back online

restart works but its looks more like workaround - not solution
